dockerfiles/Dockerfile-toolchain

```diff
@@ -39,26 +47,79 @@ RUN sudo git clone https://github.com/facebook/rocksdb -b v5.17.2 /rocksdb
 # This builds and installs just the static lib for us
 RUN cd /rocksdb && sudo make static_lib PORTABLE=1 -j$(nproc) && sudo cp librocksdb.a /usr/local/lib/librocksdb_coda.a && sudo rm -rf /rocksdb && sudo strip -S /usr/local/lib/librocksdb_coda.a
 
+# save commits for pinned packages, to know if they should be updated in CI
+RUN sudo git clone https://github.com/CodaProtocol/coda /coda
+
+# submodules
+RUN for pkg in async_kernel digestif graphql_ppx ocaml-extlib ppx_optcomp rpc_parallel ; do \
+      # remove leading "-" indicating uninitialized
+      cd /coda && sudo bash -c "git submodule status src/external/$pkg | awk '{print \$1}' | sed s/-// > ~opam/opam-repository/$pkg.commit" ; \
+    done
+
+# use log commit for ocaml-sodium, ocaml-rocksdb, coda_base58, which are not submodules
+RUN for pkg in ocaml-sodium ocaml-rocksdb coda_base58 ; do \
+      cd /coda && sudo bash -c "git log --format=oneline -n 1 src/external/$pkg | awk '{print \$1}' > ~opam/opam-repository/$pkg.commit" ; \
+    done
+
+# repo was only needed for submodule commits, remove it
+RUN sudo rm -rf /coda
+
 # OPAM dependencies
 RUN git -C /home/opam/opam-repository pull
 RUN opam update -y && opam upgrade -y
 
+
 # spacetime OCaml compiler
 RUN opam switch create $OCAML_VERSION+spacetime
 RUN eval $(opam env)
 
 # Install OPAM packages for spacetime compiler; use ulimited stack for camlp4 build
 ADD /src/opam.export .
-RUN ulimit -s unlimited && opam switch import opam.export ; rm opam.export
+RUN ulimit -s unlimited && opam switch import opam.export
+
+# Add packages to be pinned
+
+# Source copy of ocaml-sodium (modified for static linking)
+ADD /src/external/ocaml-sodium /ocaml-sodium
+# Source copy of rpc_parallel (modified for macOS support)
+ADD /src/external/rpc_parallel /rpc_parallel
+# Source copy of ocaml-extlib (modified to remove module name conflict)
+ADD /src/external/ocaml-extlib /ocaml-extlib
+# Source copy of digestif (modified to support unpadded SHA256)
+ADD /src/external/digestif /digestif
+# Source copy of mransan version of base58 (OPAM repo has two versions with different APIs)
+ADD /src/external/coda_base58 /coda_base58
+# Source copy of async_kernel with tracing support
+ADD /src/external/async_kernel /async_kernel
+# Source copy of graphql_ppx, updated for 4.07.1/4.08.0 support
+ADD /src/external/graphql_ppx /graphql_ppx
+
+# Pin packages for spacetime compiler
+
+RUN cd /ocaml-sodium && opam pin -y add .
+RUN sudo rm -f /rpc_parallel/.git && cd /rpc_parallel && opam pin -y add .
```

Comment:
I think the `rm`s should happen close to the `ADD`s to avoid inflating the intermediate docker images.